### ISSUE TITLE
Generate convenience methods for common MULTISET and ROW nestings

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#13069]
As discussed in https://github.com/jOOQ/jOOQ/issues/13063, jOOQ 3.17 will have additional convenience to build common `MULTISET` and `ROW` expressions based on path navigation, similar to the existing implicit join feature. #13063 remains a discussion issue, this issue here is a specific task list for a prototype, and then the final implementation.

The current prototype will add:

- [x] `<O extends Record> CHILD.parentRow(Function<? super Parent, ? extends TableLike<O>>)` produces a `Row[N]` which extends `SelectField<R>`
- [x] `<O extends Record> PARENT.childMultiset(Function<? super Parent, ? extends TableLike<O>>)` produces a `Field<Result<R>>` (we might introduce a `Multiset<R>` subtype of `Field<Result<R>>` in the future, see https://github.com/jOOQ/jOOQ/issues/12031). This works for:
  - [x] One-to-many relationships (trivial)
  - [x] Many-to-many relationships. A heuristic is implemented where for a table `T`, all child tables' `C`'s foreign keys are resolved to point to `U`, meaning `T <-> U` are likely to be in a many-to-many relationship

### Code generation

3 new `<generate/>` flags are added to allow for opting out of the feature, to prevent unnecessary noise if the feature is not desired, as well as to work around code generation bugs. All flags default to true:

- [ ] Generate code only if `MULTISET` emulations are available
- [x] `<existsConvenience/>`? Could be useful as well, to be defined
- [x] `<rowConvenienceToOne/>`
  - [x] A prerequisite is to fix this bug: https://github.com/jOOQ/jOOQ/issues/12930
- [x] `<multisetConvenienceOneToMany/>`
- [x] `<multisetConvenienceManyToMany/>`
  - [ ] Special case: There are `UNIQUE` constraints on the relationship table's FKs: https://github.com/jOOQ/jOOQ/issues/13074
  - [ ] Special case: There are multiple paths between two many-to-many connected tables
  - [ ] Exclude case: There are more than 2 FKs on the relationship table
  - [x] Exclude case: There isn't any key on the 2 FKs
  - [x] Exclude case: A FK column is nullable
  - [ ] An option could be added to exclude the relationship table from the one-to-many convenience

All of these features must be made available to:

- [x] `JavaGenerator`
- [x] `KotlinGenerator`
- [x] `ScalaGenerator`

### Generator Strategy

Just like for implicit join paths, the naming depends on the foreign key name, but this time in the inverse direction:

- [ ] We must be able to define names based on inverse keys (i.e. using the referencing table, not the referenced table)
  - [ ] In case of disambiguation, the foreign key name is still used
  - [ ] Users should be able to implement a custom naming strategy as well
- [x] The general naming needs to be specified. The suggested default is to suffix the key name (or table name, if non-ambiguous) with `Multiset` or `Row`, but that might change.

### Prerequisites

- [ ] https://github.com/jOOQ/jOOQ/issues/8012. This would lead to much more convenience in the lambda syntax
- [ ] https://github.com/jOOQ/jOOQ/issues/12031. (The `Multiset<R>` type seems desirable)
- [ ] https://github.com/jOOQ/jOOQ/issues/13065.
- [ ] https://github.com/jOOQ/jOOQ/issues/12045. (Would be great!)
- [ ] https://github.com/jOOQ/jOOQ/issues/12772. (We need `QueryPart::$replace` to traverse join trees to get this right)

### Related

- https://github.com/jOOQ/jOOQ/issues/4727
- https://github.com/jOOQ/jOOQ/issues/13066
- https://github.com/jOOQ/jOOQ/issues/13067

### Tests and feature interactions

- [x] Aliasing of tables
- [x] Aliasing of the multiset expression
- [x] Implicit joins from within the `MULTISET` subquery
- [ ] Self joins for tree structures
- [ ] Converters and bindings
- [ ] Composite keys
- [ ] Embeddable keys

**Comment by lukaseder** [comment]
Some convenience to generate correlated `EXISTS` subqueries could be useful as well to simplify things like:

```sql
SELECT
  t.a,
  t.b,
  EXISTS (
    SELECT 1
    FROM u
    WHERE u.t_id = t.t_id
  )
FROM t
```

E.g. on `T_LANGUAGE`:

- `Condition tBookExists()`
- `Condition tBookExists(Function<? super TBook, ? extends TableLike<O>> subquery)`

**Comment by lukaseder** [comment]
A related improvement: https://github.com/jOOQ/jOOQ/issues/13552

**Comment by morki** [comment]
I tried those new functions but without success with mapping the resulting types. Probably it is just misunderstanding of the concept.

For example I thought that:

```kotlin
T_BOOK.authorRow { author -> select(row(author.NAME, author.AGE).mapping(::Author)).from(author)  }
```

Should be equivalent to:

```kotlin
row(
  T_BOOK.author().NAME,
  T_BOOK.author().AGE,
).mapping(::Author)
```

But it is not. The first returns `Field<Record1<Author>>` instead of `SelectField<Author>` which is returned by the second one.

**Comment by lukaseder** [comment]
You probably mean this?

```kotlin
T_BOOK.authorRow { a -> select(a.NAME, a.AGE).from(a) }.mapping(::Author)
```

**Comment by lukaseder** [comment]
Ah, no that extension function `Field<Row[N]>.mapping(...)` does not yet exists, it probably should. It's this, then:

```kotlin
T_BOOK.authorRow { a -> select(a.NAME, a.AGE).from(a) }.convertFrom { it.map(mapping(::Author)) }
```

Not really that much better than the implicit join version in this particular case

**Comment by morki** [comment]
Yes, this was the problem why I transfer the mapping into the lambda. Maybe adding this `mapping` function to `Field<...>` will make it clear for other people too.

**Comment by lukaseder** [comment]
It can't be added to `Field` itself. It can only exist as an extension function

**Comment by lukaseder** [comment]
No worries. I've pushed a commit here: https://github.com/jOOQ/jOOQ/issues/13634, see https://github.com/jOOQ/jOOQ/commit/c98f4cf6f78ebaebf910d72cf984c86fbd0e7c7e

**Comment by lukaseder** [comment]
> Another idea when I test this new feature is possibility to optionally merge the declaration with mapping like this:

This is mainly a feature of the Java API, so I think this kotlin version isn't feasible? Besides, I don't like inversing the concerns. The main feature here is to project something to become a jOOQ structural type (e.g. `Field<Record2<String, Integer>>`). The fact that this can be mapped to some user type is opt-in, and shouldn't precede the more important query logic.

**Comment by lukaseder** [comment]
Anyway, apart from #13634 being useful independently (users might write `field(select(row(A, B)))` and wish to map that to a custom type of degree 2 just the same), what are your thought regarding this feature *in general*?

Personally, I'm not too happy yet. While the general direction is good, the details haven't matured yet. Here's a summary:

### Naming

The naming is probably reasonable for the to-one case. It follows the implicit join path naming, hinting at a foreign key being navigated

- Caveat: One-to-many inverse navigation (from parent to children) still needs some thought regarding naming, because the parent doesn't "own" the name, the children do
- Caveat: Many-to-many navigation (from parent to parent via relationship table) needs even more thought regarding naming. Apart from using the target able name if the path is unambiguous, there isn't really a very good default, as 2 foreign keys are involved

While naming might be a detail, it's key to success of this feature as naming and convenience are very closely related.

### Idiomacy

The whole concept is quite novel in jOOQ, namely:

- Using a lambda expression to produce a subquery
- Having the lambda argument be an internal `Table` type that filters on a table
  - The current implementation hints at the complexities. We're currently passing the table itself, adding the predicate only after the lambda completes.
  - We do that because there are still too many irregularities in the existing `Table::where` implementations (including the fact that generated code does not yet override it: https://github.com/jOOQ/jOOQ/issues/8012)
- The lambda signature doesn't clearly document what's expected from the result. It makes sense after some thought, but I can't get rid of the feeling that it is *very* weird.
- It only really works on generated code, not on arbitrary expressions (just like implicit joins).

I'm always reluctant to introduce such new concepts because it's hard to assess whether they scale in terms of API design. If this design can't be reused elsewhere, it's not a good design. Foreign key path navigation seems to be an area where additional restrictions can be provided by a lambda. What else?

### Examples

Taking it a step back, the feature's main purpose is to simplify common operations. It doesn't really add any fundamental functionality, only convenience.

Compared to implicit joins, which also don't add fundamental functionality, only convenience, these "convenience methods" (better name might be defined) don't feel powerful enough. I've tried playing around with them in the new demo: https://github.com/jOOQ/demo (which is using the sakila database: https://www.jooq.org/sakila). Any time I tried to do something sophisticated, they weren't good enough. Examples:

**Using `multiset` convenience**:

- Navigating from `FILM` all the way to `PAYMENT` is too complicated. Much simpler with ordinary joins / implicit joins
- Aggregating stuff doesn't work as well either, it doesn't feel right

**Using `row` convenience**:

- That one was mainly added for consistency reasons (to have the same functionality in both directions of the foreign key navigation)
- It doesn't seem very useful
- In particular, as @morki's examples have shown here: https://github.com/jOOQ/jOOQ/issues/13069#issuecomment-1145226775, it's not really better than just using the primitives, namely `row(book.author().ID, book.author().NAME)`. I guess this is about avoiding to repeat the paths, but that's not a strong use-case. The path could be assigned to a local variable, too.

**Using `exists` convenience**:

- It seemed to make sense to "aggregate" the `multiset` into a single boolean value
- Again, it was very hard to get the right result out of the `EXISTS` check, once I wanted to navigate further than just the immediate foreign key
- This particular convenience method suggests that we might need other convenience, too? E.g. for `COUNT(*)`? But where does it stop?

### Conclusion

- The underlying primitives (`MULTISET`, `ROW`) are sound and *very* useful. They fit right in the SQL language, being part of the SQL standard too, and having some successful implementations in native form or in emulated forms.
- They *compose* very well, being very versatile operators.
- Implicit joins also feel very natural and useful.
- These convenience methods seem clunky. Some very simple use-cases now become easier, but some sophisticated use-cases become harder, making it difficult for developers to assess whether they should shoehorn their subquery into this new feature, or revert to the primitives (`MULTISET`, `ROW`, etc.)

Clearly, as it is now, this feature can't be taken out of experimental state. I'll even have to turn it off by default, I think, so only power users will enable it and play with it. I wouldn't want to suggest this feature will remain as it is, so people won't complain (too much), when there are breaking changes in 3.18, etc.

I'd like to ship jOOQ 3.17 soon, so I think this area won't see much improvement anymore for this release. I want to invest more time into solving the *underlying* problems, so maybe, we should take a step back and discuss again what problem this issue was supposed to solve (the discussion issue is here: https://github.com/jOOQ/jOOQ/issues/13063)

### Taking a step back

Maybe, this whole feature was too *"specific"*, trying to solve only one isolated problem, rather than adding dozens of little improvements, and thinking of the big picture. For example, it seems clear that these two primitives are very useful and should be improved:

- `Table::where` (existing, improvements will be interesting)
- `Table::select` (not yet existing: https://github.com/jOOQ/jOOQ/issues/13066)

These two operations represent the two most fundamental operations of relational algebra: selection and projection. Having purely table based convenience for those seems reasonable. Even SQL has similar primitives. In standard SQL and PostgreSQL, you can write `TABLE book ORDER BY id LIMIT 3`. The available primitives are different. It's hardly ever useful other than just quickly checking something in ad-hoc SQL use-cases, and even then, it just shortens `SELECT * FROM book` to `TABLE book`. Not a big win. But together, both jOOQ's and SQL's short versions show that occasionally, having additional means of applying `SELECT` clauses to tables directly, rather than going through the entire `SELECT` ceremony can help.

But the real problem that was originally discussed here https://github.com/jOOQ/jOOQ/issues/13063, was that *correlated subqueries are too clunky in SQL*. This isn't even about `MULTISET`. This is also about a query like this:

```sql
SELECT
  a.FIRST_NAME,
  a.LAST_NAME,
  (
    SELECT count(*) 
	FROM book
	WHERE author_id = a.id
  )
FROM author AS a
```

People tend to want something like this, instead (pseudo syntax):

```sql
SELECT
  a.FIRST_NAME,
  a.LAST_NAME,
  a.books(COUNT(*))
FROM author AS a
```

So, what people want is:

- Path navigation (of arbitrary path length) (e.g. `a.books(...)`, but also `a.books.languages(...)`
- Projection of the path target tuples (e.g. `COUNT(*)`, but also `COUNT(*), languages(COUNT(*))`
- Choice of projecting 1 (`ROW`) or N (`MULTISET`) rows (e.g. `COUNT(*)` but also `ID`)

I feel the current solution is close, yet still so far 😅. More dogfooding is needed, more feedback is greatly appreciated!

**Comment by lukaseder** [comment]
So, one thing I had been trying to do a few times now, always being reminded of the limitations of the current design. Let's assume I want to get all the *distinct* categories per actor, like this:

```java
ctx.select(
      ACTOR,
      multiset(
          selectDistinct(FILM_CATEGORY.category().NAME)
          .from(FILM_CATEGORY)
          .join(FILM_ACTOR).on(FILM_CATEGORY.FILM_ID.eq(FILM_ACTOR.FILM_ID))
          .where(FILM_ACTOR.ACTOR_ID.eq(ACTOR.ACTOR_ID))
      ))
   .from(ACTOR)
   .fetch();
```

This produces something like:

```
+-----------------------------------------------+--------------------------------------------------+
|actor                                          |multiset                                          |
+-----------------------------------------------+--------------------------------------------------+
|(81, SCARLETT, DAMON, 2006-02-15T04:34:33)     |[(Action), (Animation), (Children), (Classics),...|
|(144, ANGELA, WITHERSPOON, 2006-02-15T04:34:33)|[(Action), (Animation), (Classics), (Comedy), (...|
|(149, RUSSELL, TEMPLE, 2006-02-15T04:34:33)    |[(Animation), (Children), (Classics), (Comedy),...|
|(135, RITA, REYNOLDS, 2006-02-15T04:34:33)     |[(Animation), (Children), (Classics), (Comedy),...|
+-----------------------------------------------+--------------------------------------------------+
```

For this, we cannot use the multiset convenience methods. Currently, there is:

- `ACTOR.filmActorMultiset()` (one-to-many)
- `ACTOR.filmMultiset()` (many-to-many)

But we need to "jump" a step further, e.g. via `FILM.categoryMultiset()` to `CATEGORY.NAME`. We could do this:

```java
ctx.select(
    ACTOR,
    ACTOR.filmMultiset(f ->
        selectDistinct(FILM_CATEGORY.category().NAME)
            .from(f)
            .join(FILM_CATEGORY).on(FILM_CATEGORY.FILM_ID.eq(f.FILM_ID))
    ))
.from(ACTOR)
.fetch();
```

Apart from the fact that there's a bug in this particular implementation, that would be mixing styles, so not really what people probably want. Not a big win compared to the original query.

This seems more like it at first:

```java
ctx.select(
    ACTOR,
    ACTOR.filmMultiset(f -> select(f.categoryMultiset(c -> selectDistinct(c.NAME).from(c))).from(f)))
.from(ACTOR)
.fetch();
```

But this isn't what we wanted:

```
+-----------------------------------------------+--------------------------------------------------+
|actor                                          |multiset                                          |
+-----------------------------------------------+--------------------------------------------------+
|(81, SCARLETT, DAMON, 2006-02-15T04:34:33)     |[([(Comedy)]), ([(Games)]), ([(Drama)]), ([(New...|
|(144, ANGELA, WITHERSPOON, 2006-02-15T04:34:33)|[([(Sports)]), ([(Classics)]), ([(Sports)]), ([...|
|(149, RUSSELL, TEMPLE, 2006-02-15T04:34:33)    |[([(Classics)]), ([(Travel)]), ([(Comedy)]), ([...|
|(135, RITA, REYNOLDS, 2006-02-15T04:34:33)     |[([(Children)]), ([(New)]), ([(Travel)]), ([(Co...|
|(115, HARRISON, BALE, 2006-02-15T04:34:33)     |[([(Sci-Fi)]), ([(Sci-Fi)]), ([(Horror)]), ([(D...|
+-----------------------------------------------+--------------------------------------------------+
```

Because:

- It produces one layer of nesting too many
- It doesn't produce distinct category names *per `ACTOR`* but *per `FILM` per actor* (see the last entry with doubled `Sci-Fi` entries.

So, what we really wanted is some sort of path navigation, similar to implicit joins (which follow to-one relationships):

```java
ctx.select(
    ACTOR,
    ACTOR.films().categoryMultiset(c -> selectDistinct(c.NAME).from(c)))
.from(ACTOR)
.fetch();
```

Similar ideas have been discussed in the original implicit join feature request, and then a follow up task:

- https://github.com/jOOQ/jOOQ/issues/1502
- https://github.com/jOOQ/jOOQ/issues/12037

So, what's really want here is a way to produce a (derived) table following path navigation (`ACTOR` -> `FILM` -> `CATEGORY`), and when we reach the end of this navigation, we need a way to project stuff.

However, even this example is too simplistic. It's not unlikely that users will want to take along some items from the path. E.g. we might want to count the number of films per category and actor, e.g. like this:

```java
ctx.select(
    ACTOR,
    multiset(
        select(FILM_CATEGORY.category().NAME, count())
        .from(FILM_CATEGORY)
        .join(FILM_ACTOR).on(FILM_CATEGORY.FILM_ID.eq(FILM_ACTOR.FILM_ID))
        .where(FILM_ACTOR.ACTOR_ID.eq(ACTOR.ACTOR_ID))
        .groupBy(FILM_CATEGORY.CATEGORY_ID, FILM_CATEGORY.category().NAME)
    ))
.from(ACTOR)
.fetch();
```

Producing

```
+-----------------------------------------------+--------------------------------------------------+
|actor                                          |multiset                                          |
+-----------------------------------------------+--------------------------------------------------+
|(81, SCARLETT, DAMON, 2006-02-15T04:34:33)     |[(Children, 4), (Documentary, 1), (Horror, 2), ...|
|(144, ANGELA, WITHERSPOON, 2006-02-15T04:34:33)|[(Travel, 4), (Documentary, 3), (Horror, 1), (D...|
|(149, RUSSELL, TEMPLE, 2006-02-15T04:34:33)    |[(Travel, 3), (Children, 2), (Documentary, 3), ...|
|(135, RITA, REYNOLDS, 2006-02-15T04:34:33)     |[(Travel, 2), (Children, 1), (Documentary, 1), ...|
|(115, HARRISON, BALE, 2006-02-15T04:34:33)     |[(Children, 1), (Horror, 3), (Documentary, 2), ...|
+-----------------------------------------------+--------------------------------------------------+
```

(One of my favourite examples for a while: https://twitter.com/lukaseder/status/832939068933681152)

We can see that `SCARLETT DAMON` played in:

- 4 Children films
- 1 Documentary
- 2 Horror films
- etc.

It seems possible to implement this with a hypothetical syntax like this:

```java
ctx.select(
    ACTOR,
    ACTOR.films().categoryMultiset(c -> 
        select(c.NAME, count())
        .from(c)
        .groupBy(c.CATEGORY_ID, c.NAME)
    ))
.from(ACTOR)
.fetch();
```

But what if we wanted to list the films with `listAgg()`? Easy with ordinary `MULTISET`:

```
+-----------------------------------------------+--------------------------------------------------+
|actor                                          |multiset                                          |
+-----------------------------------------------+--------------------------------------------------+
|(81, SCARLETT, DAMON, 2006-02-15T04:34:33)     |[(Action, CROW GREASE, MINDS TRUMAN), (Animatio...|
|(144, ANGELA, WITHERSPOON, 2006-02-15T04:34:33)|[(Action, BERETS AGENT, BULL SHAWSHANK, CELEBRI...|
|(149, RUSSELL, TEMPLE, 2006-02-15T04:34:33)    |[(Animation, CANYON STOCK, CROSSROADS CASUALTIE...|
|(135, RITA, REYNOLDS, 2006-02-15T04:34:33)     |[(Animation, LUKE MUMMY), (Children, MAKER GABL...|
|(115, HARRISON, BALE, 2006-02-15T04:34:33)     |[(Animation, DOUBLE WRATH, NASH CHOCOLAT, PUNK ...|
+-----------------------------------------------+--------------------------------------------------+
```

But I can't see how we can reach the `FILM` information again in the multiset convenience approach:

```java
ACTOR.films().categoryMultiset(c -> ???)
```

We can still do it if we mix styles, though:

```java
ctx.select(
        ACTOR,
        ACTOR.filmMultiset(f ->
           select(
               FILM_CATEGORY.category().NAME,
               listAgg(f.TITLE, ", ").withinGroupOrderBy(f.TITLE))
           .from(f)
           .join(FILM_CATEGORY).on(FILM_CATEGORY.FILM_ID.eq(f.FILM_ID))
           .groupBy(FILM_CATEGORY.CATEGORY_ID, FILM_CATEGORY.category().NAME)
        ))
    .from(ACTOR)
    .fetch();
```

(The result is again wrong because there's still a bug)

But again, it's not really much better than the version that doesn't use this multiset convenience.

I guess we're stuck for now... This convenience feature only adds value for very simple path navigation cases.

@alf & @morki, if you could share a few real-world `MULTISET` queries (publicly or privately), I'd really appreciate it!

**Comment by alf-lervag** [comment]
I’ll try to get back to you with more on this one next week. Here’s the shortsighted answer based on what’s at the front of my mind right now. 

We’re going to be implementing a lot of GraphQL-stuff. I believe this means that:

1. Implicit joins will make our lives way easier when it comes to fetching related data, both multisets and rows
2. We generally don’t need to support aggregations

**Comment by alf** [comment]
I understand the problem and I’m happy that you’re trying to solve the general case. I will do my best to assist with use cases and input.

GraphQL does not aim to solve these issues, but it does provide value to us. We need the quick wins. 🤷‍♂️

That said, we’re in this for the long run and trends change. SQL not so much. Thus the goals of jOOQ aligns very well with what we believe. Let’s get this right.

**Comment by lukaseder** [comment]
> GraphQL does not aim to solve these issues, but it does provide value to us. We need the quick wins. 🤷‍♂️

Makes perfect sense, @alf

**Comment by lukaseder** [comment]
So, perhaps, let's take again inspiration in HQL, which already inspired implicit joins. Going back to the thing that is being improved here: simplifying correlated subquerying, irrespective of `MULTISET`. Instead of:

```sql
SELECT
  a.FIRST_NAME,
  a.LAST_NAME,
  (
    SELECT count(*) 
    FROM book
    WHERE author_id = a.id
  )
FROM author AS a;
```

HQL suggests we should be able to write this, instead:


```sql
SELECT
  a.FIRST_NAME,
  a.LAST_NAME,
  (
    SELECT count(*) 
    FROM a.book -- Implicit join or correlation, depending on context
  )
FROM author AS a;
```

> **Note**: *I'm not sure if HQL actually supports this particular usage. I've found a few flaws in their understanding of implicit joins and how it relates to producing sets / cartesian products / etc. For example, even their to-one implicit joins always produce (or used to?) inner joins, leading to an implicit join present only in the `SELECT` clause potentially removing rows from the result set, in case the foreign key is optional. I'm guessing their to-many implicit joins may have similar issues, to be investigated*

So, perhaps, the sophisticated `MULTISET` example I made before:

```java
ctx.select(
    ACTOR,
    multiset(
        select(
            FILM_CATEGORY.category().NAME, 
            listAgg(FILM_CATEGORY.film().TITLE, ", ")
                .withinGroupOrderBy(FILM_CATEGORY.film().TITLE))
        .from(FILM_CATEGORY)
        .join(FILM_ACTOR).on(FILM_CATEGORY.FILM_ID.eq(FILM_ACTOR.FILM_ID))
        .where(FILM_ACTOR.ACTOR_ID.eq(ACTOR.ACTOR_ID))
        .groupBy(FILM_CATEGORY.CATEGORY_ID, FILM_CATEGORY.category().NAME)
    ))
.from(ACTOR)
.fetch();
```

Would really translate to this:

```java
ctx.select(
    ACTOR,
    multiset(
        select(
            ACTOR.film().category().NAME, 
            listAgg(ACTOR.film().TITLE, ", ")
                .withinGroupOrderBy(ACTOR.film().TITLE))
        .from(ACTOR.film(), ACTOR.film().category())
        .groupBy(
            ACTOR.film().category().CATEGORY_ID, 
            ACTOR.film().category().NAME)
    ))
.from(ACTOR)
.fetch();
```

It would be totally possible to use aliases, too!

```java
Film f = ACTOR.film().as("f");
Category c = f.category().as("c");

ctx.select(
    ACTOR,
    multiset(
        select(c.NAME, listAgg(f.TITLE, ", ")
            .withinGroupOrderBy(f.TITLE))
        .from(f, c)
        .groupBy(c.CATEGORY_ID, c.NAME)
    ))
.from(ACTOR)
.fetch();
```

Or, again, with the simpler, initial example:

```java
ctx.select(
      ACTOR,
      multiset(
          selectDistinct(FILM_CATEGORY.category().NAME)
          .from(FILM_CATEGORY)
          .join(FILM_ACTOR).on(FILM_CATEGORY.FILM_ID.eq(FILM_ACTOR.FILM_ID))
          .where(FILM_ACTOR.ACTOR_ID.eq(ACTOR.ACTOR_ID))
      ))
   .from(ACTOR)
   .fetch();
```

We'd get, instead:

```java
ctx.select(
      ACTOR,
      multiset(
          selectDistinct(ACTOR.film().category().NAME)
          .from(ACTOR.film().category())
      )
   .from(ACTOR)
   .fetch();
```

The `EXISTS` case works just as fine, e.g. finding all `CATEGORY` and checking if there exists any `ACTOR` that played in any `FILM` of that `CATEGORY`:

```java
ctx.select(CATEGORY)
   .from(CATEGORY)
   .where(exists(selectOne().from(CATEGORY.film().actor())))
   .fetch();
```

The `ROW` case wouldn't need a new representation in this world. @morki's example would suffice:

```kotlin
row(
  BOOK.author().NAME,
  BOOK.author().AGE,
)
```

The details would have to be fleshed out. So far, I've rejected this kind of to-many implicit join because it has its own caveats:

```java
ctx.select(
        ACTOR.FIRST_NAME,
        ACTOR.LAST_NAME,

        // This innocent looking projection seems to
        // produce a cartesian product
        ACTOR.film().TITLE)
   .from(ACTOR)
   .fetch();
```

I.e. when supporting navigational to-many implicit joins based on paths, it's possible for non-`FROM` clauses to produce cartesian products, which is very un-SQL-esque. Imagine a `WHERE` clause that doesn't filter out rows, but instead creates new ones!

Without having thought this through, we could add a requirement for any such paths to be present in the `FROM` clause, e.g. the above would be illegal, but this would work:

```java
ctx.select(
        ACTOR.FIRST_NAME,
        ACTOR.LAST_NAME,

        // Now, the cartesian product is made explicit in
		// FROM, so it's no longer SELECT's fault
        ACTOR.film().TITLE)
   .from(ACTOR, ACTOR.film())
   .fetch();
```

The hypothetical rule being:

- All `to-many` implicit join paths must be present in the from clause explicitly
- All `to-one` implicit join paths may or may not be present in the from clause explicitly

### Conclusion

This approach solves the same problem: Simplified correlated subqueries. But it solves it in a much more SQL-idiomatic way, i.e. in a way that might as well be available in native SQL. It does not introduce any new idioms via lambdas and implied correlation on the lambda arguments. It does not just address the simple navigational patterns of GraphQL-style SQL usage, but also helps simplify the tricky edge cases, by simplifying both:

- Correlated subqueries
- Arbitrary join expressions

The feature could even obsolete the current `onKey()` clause.

It's Friday night, so I guess this won't find a solution again tonight, but I'm confident we're closing in on the right solution :)

**Comment by alf** [comment]
I like where this is going. Well done!

**Comment by alf** [comment]
One benefit of `onKey` is that it allows me to add extra conditions on the join. That’s probably not very important but I figured I’d mention it.

**Comment by lukaseder** [comment]
I just used table lists for brevity. It should be possible to use implicit joins also with the `JOIN` syntax. It won't be trivial to implement, but it should be straightforward to use.

**Comment by lukaseder** [comment]
For the record, by "obsolete," I don't mean the feature will be removed. I just mean it will probably be used much less.

**Comment by morki** [comment]
I was playing with the prototype a bit last days and I must say I aggree with you @lukaseder. From the first sight I was impressed how "simple" it is to use this and it was mentally closer to the DTO model. 

But after a moment and some "harder" use cases, it feels ugly, not understandable, not directly "translatable" to SQL. It reminded me of Hibernate, where now I need to have 3 different models in mind - SQL, jOOQ, DTO. I does not really simplify things.

I even don't like your next try, it seems super magical and more unmaintainable for me, even with the requirement of duplicating in FROM clause.

But thank you very much for this experiment, it was fun to play with it and I make the final decision to pay the PRO edition :)

**Comment by lukaseder** [comment]
> But after a moment and some "harder" use cases, it feels ugly, not understandable, not directly "translatable" to SQL. It reminded me of Hibernate, where now I need to have 3 different models in mind - SQL, jOOQ, DTO. I does not really simplify things.

So, I did correctly interpret the tumbleweeds in this conversation 😅.

I'll review this a bit more next week and probably even remove the prototype again. If something is this flawed and feels this wrong, it shouldn't be kept around for too long. Once this is released to 3.17 (even as an "experimental" prototype), people will start using it, and it will be much harder to change / remove again.

> I even don't like your next try, it seems super magical and more unmaintainable for me, even with the requirement of duplicating in FROM clause.

I'm very curious about that if you have some more time. Why does it feel magical and unmaintainable? (I'm assuming from a user perspective? I don't see significant issues from an implementation perspective). Why is it different from today's implicit joins (the to-one version)? Or do you feel the same about today's implicit joins?

Maybe, some details in design need to be fleshed out, but I feel that this works very much like standard SQL `LATERAL` or T-SQL/Oracle `APPLY`:

```sql
-- Valid in Oracle, SQL Server today
SELECT *
FROM book 
CROSS APPLY (
  SELECT * FROM author WHERE author.id = book.author_id
)

-- Valid in Db2, Oracle, PostgreSQL today
SELECT *
FROM book 
CROSS JOIN LATERAL (
  SELECT * FROM author WHERE author.id = book.author_id
)
```

Now, people have created table valued functions to simplify such `APPLY` usage in all RDBMS that support the syntax, to replace the above with this:

```sql
SELECT *
FROM book 
CROSS APPLY authors(book.author_id)
```

This is particularly popular in SQL Server, which has supported powerful inline table valued functions for ages (meaning the function is transparent to the optimiser, it can be inlined into the calling SQL query).

Some RDBMS even allow implicit `LATERAL` / `APPLY` for historic reasons, as the `LATERAL` keyword hasn't been standardised earlier (e.g. Db2, Oracle, PostgreSQL):

```sql
SELECT *
FROM book, authors(book.author_id)
```

In older Db2, Oracle, a `TABLE` operator was needed, or may still be needed in edge cases:

```sql
SELECT *
FROM book, TABLE(authors(book.author_id))
```

At which point the slight change of notation seems tempting?

```sql
SELECT *
FROM book, book.authors -- instead of authors(book)
```

And once that notation is available, it can either be reused in other clauses:

```sql
SELECT book.*, book.authors.*
FROM book, book.authors
```

Or it can be aliased:


```sql
SELECT b.*, a.*
FROM book AS b, book.authors AS a
``` 

And now, we have pretty much the semantics HQL has introduced for Hibernate, except that Hibernate doesn't also support `LATERAL` and all the rest, as far as I know. They seem to have built the same intuition as early Db2, Oracle, PostgreSQL, though, that this kind of "lateral" path navigation is logical in the `FROM` clause.

Again, the only reason why I had rejected it in jOOQ 3.11 (when implicit joins were introduced), is because there are still a few open questions with producing "surprise cartesian products" from arbitrary clauses, but once the table has to be declared in `FROM`, there's no more surprise, I think?

> But thank you very much for this experiment, it was fun to play with it and I make the final decision to pay the PRO edition :)

At least, we got something out of all this :)

**Comment by morki** [comment]
I really didn't know any of these techniques, so it was feeling magical for be, because I couldn't translate it in my head into some similar SQL. With your introduction, it feels pretty ok now :)

**Comment by lukaseder** [comment]
If anything at all, your feedback is a stark reminder to make sure the right "story" is told in the manual, and in some tutorials, rather than leaving users to discover how to connect the dots themselves...

This is also true for the existing `LATERAL` / `APPLY` / table valued functions support. From my past SQL trainings, I've often noticed that this feature set "clicks" immediately with half of the audience, but not with the other half. I'm still looking for better ways to explain these things... I often make a `Stream.flatMap()` analogy, because that's what `APPLY` really is, e.g.

```java
Stream<Tuple2<Book, Author>> stream =
books.stream()
     .flatMap(b -> b.authors().map(a -> tuple(b, a)));
```

**Comment by alf** [comment]
I did a workshop this week where I tried to show how powerful jOOQ is for doing dynamic SQL.

I find that being able to extract and re-use projections is fantastic. Especially since we can map to user types here. For instance we can do stuff such as:

```
ctx.select(createPerson(PERSON))
   .from(PERSON)
   .where(…)
```

and then:

```
ctx.select(createPerson(ASSIGNMENT.person())))
   .from(ASSIGNMENT)
   .where(…)
```

I haven’t tested this completely, but from what I’ve seen the createPerson-function can do implicit joins in addition to mapping to user types.

If that holds, I believe the new design you hint at here would also allow this pattern to do multiset-joins in the projection quite naturally.

**Comment by lukaseder** [comment]
Thanks for that feedback, @alf. Yes, that's precisely one of the powerful use-cases of using jOOQ's dynamic SQL capabilities: To be able to re-use type safe projections across queries.

> I haven’t tested this completely, but from what I’ve seen the createPerson-function can do implicit joins in addition to mapping to user types.

The `createPerson()` function can indeed do implicit joins. Think of `select()` like a `Stream.map(...)` operation, which is the `Stream` version of the projection. You can map an input tuple to any output tuple. An implicit join is just a special projection. I used the term "implicit join," because HQL uses that term, and it kinda makes sense, because that's what the ideal *implementation* does. But it doesn't have to be a join, it could also be implemented as a correlated subquery, which is what jOOQ 3.17 does when you use such "path expressions" in DML statements: https://github.com/jOOQ/jOOQ/issues/7508 (joins are much harder to achieve in vendor agnostic DML). And a correlated subquery is just a fancy projection. In fact, any correlated subquery can be transformed to a `LEFT JOIN`, which is what some optimisers, e.g. Oracle, know to do when subqueries are similar.

> If that holds, I believe the new design you hint at here would also allow this pattern to do multiset-joins in the projection quite naturally.

To be accurate: You can use `MULTISET` correlated subqueries, yes. That's probably what you meant by "multiset-join"? They work no different from ordinary correlated subqueries, except that ordinary correlated subqueries are scalar subqueries (1 row x 1 column), whereas with `MULTISET`, you can nest collections (N rows) of tuples (N columns).

But they're not strictly joins, which are a feature of the `FROM` clause and cause cartesian products (by design). As I mentioned here: https://github.com/jOOQ/jOOQ/issues/13069#issuecomment-1146064183 with this comment:

```java
        // This innocent looking projection seems to
        // produce a cartesian product
```

You probably won't be able to project a "to-many path expression" without also declaring it in the `FROM` clause, because the `SELECT` clause must not be able to generate or filter any rows. `SELECT` = `Stream.map()`, not `Stream.flatMap()` or `Stream.filter()`. This might be a bit confusing at first, and I'll try hard to either explain this very well in manuals, etc., or find an API design way to prevent the confusion (e.g. by using a different name on the generated method, depending on whether the method traverses the path from child to parent or from parent to child).

Anyway. You will be able (I guess) to use the "to-many path expression" in a `MULTISET` correlated subquery from that `SELECT` clause.

So, this issue also shows that naming will be essential in the future:

- *"Implicit join"* means that a path expression anywhere in the query implicitly produces a `JOIN` (or `LEFT JOIN`) in the appropriate `FROM` clause. This is an implementation detail that leaked into the name of the feature. The actual feature is a "path expression" (or whatever name), which can be implemented in several ways, including "implicit join" or correlated subquery. The implicit join implementation is more efficient, because it allows for sharing path segments across path expressions, whereas a correlated subquery (being scalar) only allows for one column to be projected *per subquery*, meaning that subqueries have to be repeated per shared path segment.
- *"Path expression"* (or whatever the name will be) is an expression going from `child.parent.column` (current feature) or `parent.child.column` (potential new feature). The latter requires explicit declaration in the `FROM` clause, the former can be explicit (not yet, but will be possible) or implicit (thus the current name)

### Off-topic

By the way, your example is an interesting one. The parent `PERSON` probably has multiple children, such as `ASSIGNMENT`. It might be interesting to extract the child type into a new type, such as `PersonChild`, and then declare that `person()` method on there, if possible. There are caveats to this in case of ambiguity, but this could be an interesting new feature in the future: https://github.com/jOOQ/jOOQ/issues/13638

**Comment by alf** [comment]
Thanks for clearing up the terms here. Will follow the new issue.

**Comment by lukaseder** [comment]
Folks, @alf & @morki. I was just wondering, isn't the annoying thing, namely creating correlation `Condition` expressions, most simply resolved by generating methods like:

```java
// CUSTOMER.ADDRESS_ID.eq(ADDRESS.ADDRESS_ID)
Condition c1 = CUSTOMER.eqAddress();

// CUSTOMER.ADDRESS_ID.eq(a.ADDRESS_ID) (possibly aliased, but still type safe)
Condition c2 = CUSTOMER.eqAddress(a);
```

I'm not happy with the naming, it's a bit weird. But the idea seems kinda obvious, no?

**Comment by alf** [comment]
To be honest, I don’t quite follow. I suspect that you’re onto something, but you need to spell it out a bit more. My apologies.

**Comment by morki** [comment]
The same with me, I don't get it.

**Comment by lukaseder** [comment]
Yeah nevermind then... 😅
